File: server/app-settings.js
import { db } from './db.js';

/**
 * Application preferences.
 *
 * These live in `app_settings`, the key/value table that already exists for the
 * VAPID keypair, rather than in a typed table of their own: they are individual
 * flags with no shared structure, and a column per preference would mean a
 * migration per preference.
 *
 * Notification preferences deliberately stay where they are. They have a fixed
 * set of related fields that the sweep reads together, which is what earns them
 * `notification_settings`.
 *
 * Values are stored as '0' / '1' strings, because the column is TEXT NOT NULL,
 * and read back as numbers so callers see the same shape the notification
 * settings use.
 */

const DEFAULTS = {
  /**
   * Whether the user's own writing about a company is sent to Gemini along
   * with the financial data.
   *
   * Off by default, and that is the whole point. `buildComprehensivePayload`
   * transmits conviction, target buy price, core rationale and — most
   * personally — the pre-committed sell guardrails, which are a record of what
   * would make someone abandon a position. Doc 13 §1 asked for an explicit,
   * default-off toggle rather than a line in a privacy policy. A default that
   * leaks is not a default.
   */
  ai_include_notes: 0
};

export const APP_SETTING_KEYS = Object.keys(DEFAULTS);

/** All preferences, with defaults filled in for anything never written. */
export function getAppSettings() {
  const settings = { ...DEFAULTS };
  try {
    const rows = db
      .prepare(
        `SELECT key, value FROM app_settings WHERE key IN (${APP_SETTING_KEYS.map(() => '?').join(', ')})`
      )
      .all(...APP_SETTING_KEYS);

    for (const row of rows) {
      settings[row.key] = Number(row.value) ? 1 : 0;
    }
  } catch (err) {
    // A preference read must never take down the page that asked for it.
    // Falling back to the defaults is safe precisely because the default for
    // anything privacy-bearing is the closed one.
    console.warn('[Settings] read warning:', err.message);
  }
  return settings;
}

/** Apply a partial update. Unknown keys are ignored rather than stored. */
export function updateAppSettings(patch = {}) {
  const stmt = db.prepare(
    'INSERT INTO app_settings (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value = excluded.value'
  );

  for (const key of APP_SETTING_KEYS) {
    if (!(key in patch)) continue;
    stmt.run(key, patch[key] ? '1' : '0');
  }

  return getAppSettings();
}

/**
 * Whether to attach the user's thesis to an AI analysis.
 *
 * A named helper rather than a property read at the call site, so that the
 * question "does this request carry personal data?" has exactly one answer in
 * the codebase and both hosts can mirror it.
 */
export function shouldIncludeNotesInAI() {
  return Boolean(getAppSettings().ai_include_notes);
}

File: android/data/src/main/kotlin/com/zandaulion/omaha/data/AppSettings.kt
package com.zandaulion.omaha.data

import androidx.room.Dao
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query

/**
 * Application preferences, mirroring `app_settings` in the PWA.
 *
 * Same table name, same key/value shape, same keys — `server/app-settings.js`
 * is the reference. That matters for one reason beyond tidiness: a preference
 * spelled differently on the two clients is a preference that silently reverts
 * to its default when a person moves between them, and the default here is the
 * closed one.
 *
 * The VAPID keypair the PWA keeps in this table has no Android counterpart, so
 * only the preferences are modelled.
 */
@Entity(tableName = "app_settings")
data class AppSettingRow(
    @PrimaryKey val key: String,
    /** '0' or '1', as text, because that is the column the PWA writes. */
    val value: String
)

@Dao
interface AppSettingsDao {
    @Query("SELECT * FROM app_settings WHERE key = :key")
    suspend fun get(key: String): AppSettingRow?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun put(row: AppSettingRow)
}

/**
 * The typed accessor. Defaults live here rather than at each call site, so
 * "what happens when nobody has chosen" has one answer.
 */
class AppSettings(private val dao: AppSettingsDao) {

    /**
     * Whether the user's own writing about a company is sent to Gemini.
     *
     * Off until someone turns it on, matching the PWA exactly. The payload
     * carries conviction, target buy price, rationale and — most personally —
     * the pre-committed sell guardrails, which are a record of what would make
     * somebody abandon a position. Doc 13 §1 asked for an explicit, default-off
     * toggle rather than a line in a privacy policy. A default that leaks is
     * not a default.
     */
    suspend fun aiIncludeNotes(): Boolean = readFlag(KEY_AI_INCLUDE_NOTES)

    suspend fun setAiIncludeNotes(enabled: Boolean) = writeFlag(KEY_AI_INCLUDE_NOTES, enabled)

    private suspend fun readFlag(key: String): Boolean =
        dao.get(key)?.value?.let { it == "1" || it.toIntOrNull()?.let { n -> n != 0 } == true } ?: false

    private suspend fun writeFlag(key: String, enabled: Boolean) =
        dao.put(AppSettingRow(key, if (enabled) "1" else "0"))

    companion object {
        const val KEY_AI_INCLUDE_NOTES = "ai_include_notes"
    }
}
